File: Task/Sum-of-squares/Nim/sum-of-squares.nim
import math, sequtils

proc sumSquares[T: SomeNumber](a: openArray[T]): T =
  sum(a.mapIt(it * it))

let a1 = [1, 2, 3, 4, 5]
echo a1, " → ", sumSquares(a1)

let a2: seq[float] = @[]
echo a2, " → ", sumSquares(a2)

File: Task/Sum-of-squares/Zen-C/sum-of-squares.c
import "std/vec.zc"

fn sum_squares<T>(vec: Vec<T>*) -> T {
    let sum: T = (T)0;
    for v in *vec { sum += v * v; }
    return sum;
}

fn main() {
    let vec = Vec<int>::new();
    println "{sum_squares<int>(&vec)}";
    vec << 2;
    vec << 3;
    vec << 5;
    println "{sum_squares<int>(&vec)}";

    let vec2 = Vec<f64>::new();
    vec2 << 2.5;
    vec2 << 3.5;
    vec2 << 5.5;
    println "{sum_squares<f64>(&vec2):g}";
}
